### PR TITLE
Add methods `#registered_types` and `#type_registered?`

### PR DESCRIPTION
**Pull request by tagomoris**
on Packer, Unpacker and Factory
This is working branch on the top of #71

### CODE REVIEW

**Comment by tagomoris** [comment]
@frsyuki Can you review this change?
Please merge into `register_type_wip` if ok. Then, I'll write more tests for packer/unpacker/factory and update API documents on that branch.

**Comment by tagomoris** [comment]
Pushed one more commit to show how use registered for ext when they registered.
@frsyuki Please take a look at this change.

File: ext/msgpack/factory_class.c [hunk 1]
```diff
@@ -187,6 +200,7 @@ void MessagePack_Factory_module_init(VALUE mMessagePack)
     rb_define_method(cMessagePack_Factory, "packer", MessagePack_Factory_packer, -1);
     rb_define_method(cMessagePack_Factory, "unpacker", MessagePack_Factory_unpacker, -1);
 
+    rb_define_method(cMessagePack_Factory, "registered_types_internal", Factory_registered_types_internal, 0);
```
Comment (frsyuki) [inline ext/msgpack/factory_class.c]: IMO `rb_define_private_method` is better.

**Comment by frsyuki** [comment]
Looks good :+1: 
Caching result of `msgpack_packer_ext_registry_lookup` sounds good idea. Great job.
Potential problems are:

* If class definition is changed after caching, msgpack_packer_ext_registry_lookup returns unexpected result.
* If application defines classes dynamically, size of cache increases infinitely

But I think those cases are rare. We can ignore them at least for now.

File: ext/msgpack/factory_class.c (hunk 1 above)
Comment (tagomoris) [inline ext/msgpack/factory_class.c]: Exactly.

**Comment by tagomoris** [comment]
I understand your concern, but we have a way to clear cache to enforce users to call `register_type` if they make any changes on class definitions. It should be written in documents.

**Pull request merged**